source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,363 @@ namespace EnvoyDefault {
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  std::uint64_t int_value{};
+  auto result = std::from_chars(value_string_view.begin(), value_string_view.end(), int_value);
+  if (result.ec != std::errc() || result.ptr != value_string_view.end()) {
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateHostHeader(const HeaderString& value) {
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-7.2,
+  // and RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2:
+  //
+  // Host       = uri-host [ ":" port ]
+  // uri-host   = IP-literal / IPv4address / reg-name
+  const auto host = value.getStringView();
+  if (host.empty()) {
+    return {HeaderEntryValidationResult::Action::Reject, UhvResponseCodeDetail::get().InvalidHost};
+  }
+
+  // Check if the host/:authority contains the deprecated userinfo component. This is based on RFC
+  // 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-4.2.4:
+  //
+  // Before making use of an "http" or "https" URI reference received from an untrusted source, a
+  // recipient SHOULD parse for userinfo and treat its presence as an error; it is likely being
+  // used to obscure the authority for the sake of phishing attacks.
+  auto user_info_delimiter = host.find('@');
+  if (user_info_delimiter != absl::string_view::npos) {
+    // :authority cannot contain user info, reject the header
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidHostDeprecatedUserInfo};
+  }
+
+  // Determine if the host is in IPv4, reg-name, or IPv6 form.
+  auto result = host.at(0) == '[' ? validateHostHeaderIPv6(host) : validateHostHeaderRegName(host);
+  if (!result.ok()) {
+    return result;
+  }
+
+  const auto port_string = result.details();
+  if (!port_string.empty()) {
+    // Validate the port, which will be in the form of ":<uint16_t>"
+    bool is_valid = true;
+    if (port_string.at(0) != ':') {
+      // The port must begin with ":"
+      is_valid = false;
+    } else {
+      // parse the port number
+      std::uint16_t port_int{};
+      auto result = std::from_chars(std::next(port_string.begin()), port_string.end(), port_int);
+
+      if (result.ec != std::errc() || result.ptr != port_string.end() || port_int == 0) {
+        is_valid = false;
+      }
+    }
+
+    if (!is_valid) {
+      return {HeaderEntryValidationResult::Action::Reject,
+              UhvResponseCodeDetail::get().InvalidHost};
+    }
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
```

Comment:
Note that `HeaderEntryValidationResult` copies the port number into the `std::string` While normally it would not be a problem, I think it is an abuse vector by sending requests with a hostname with very large invalid port string that would cause extensive copying in this place.

I think it is worthwhile to create a new type with the bool and abs::string_view pair that reports both validity of hostname and string view to the port number to avoid copying it.

Reply:
Done -- I added a new `HostHeaderValidationResult` class that stores the port as a `string_view`.